### PR TITLE
Verbose option

### PR DESCRIPTION
**Pull request by samuel-massinon**
Adding `verbose=true` to most functions will output libCURL's verbose messages.

@omus ~~Ready to review~~
I'm going to look into the build errors first.
Finally ready to review

### CODE REVIEW

**Comment by samuel-massinon** [comment]
@omus What’s you opinion on how to test that verbose is working? Should I just capture in output and make sure that it isn't empty or should I make sure the right message is displayed? I'm leaning to the former because I think the message is dependant on libCURL and with the online builds, I have no control over that.

File: src/FTPC.jl [hunk 1]
```diff
@@ -251,7 +251,9 @@ Download file with persistent connection.
 
 returns resp::Response
 """ ->
-function ftp_get(ctxt::ConnContext, file_name::AbstractString, save_path::AbstractString=""; mode::FTP_MODE=binary_mode)
+function ftp_get(ctxt::ConnContext, file_name::AbstractString, save_path::AbstractString=""; mode::FTP_MODE=binary_mode, verbose::Bool=false)
+    ftp_set_verboes!(ctxt, verbose)
```
Comment (omus) [inline src/FTPC.jl]: Typo

**Comment by omus** [comment]
Since you have no control over what the verbose messages are then you can just check you are getting more back than before. If you decide to include your own verbose information from FTPClient.jl you should test for that output specifically

**Comment by samuel-massinon** [comment]
Sounds good. We add nothing to the verbose messages.

**Comment by omus** [comment]
It seems like everything is getting passed in a `verbose` and `verbose_file` keyword. Maybe this would be cleaner using globals?

**Comment by samuel-massinon** [comment]
@omus Are you trying to cut down on the number of places `verbose` and `verbose_file` is passed as a function parameter? Because ever function except for 1 can be called by a user, so I don't think I can remove them.

**Comment by omus** [comment]
That's my point. Having to pass the `verbose_file` into every single FTPClient function is going to get annoying. Maybe having a `FTPClient.set_verbose(true)`that sets a global variable may be an easier option.

**Comment by samuel-massinon** [comment]
Hmm, I think my only issue with that is that to use `verbose_file` I need to create a `FILE` and close it. So if I were to do it that way I think I would have to wrap everything so I can insure the file would close.

File: src/FTPC.jl [hunk 2]
```diff
@@ -480,6 +462,41 @@ end
...
+        @ce_curl curl_easy_setopt CURLOPT_VERBOSE Int64(0)
+    end
+
+    try
+        @ce_curl curl_easy_perform
+        process_response(ctxt, resp)
+    finally
+        if isa(libc_file, Libc.FILE)
+            close(libc_file)
+            @ce_curl curl_easy_setopt CURLOPT_STDERR Libc.FILE(RawFD(2), "w").ptr
```
Comment (omus) [inline src/FTPC.jl]: Try `Libc.FILE(STDERR.handle)` here. Not entirely sure it'll work but it definitely reads better.

File: src/FTPC.jl [hunk 3]
```diff
@@ -480,6 +462,41 @@ end
...
+    else
+        @ce_curl curl_easy_setopt CURLOPT_VERBOSE Int64(0)
+    end
+
+    try
+        @ce_curl curl_easy_perform
+        process_response(ctxt, resp)
+    finally
+        if isa(libc_file, Libc.FILE)
+            close(libc_file)
```
Comment (omus) [inline src/FTPC.jl]: I'm quite worried about this. When you close the `libc_file` can you still write to the `IOStream` that `verbose_file` points to?

**Comment by omus** [comment]
I think you could definitely combine `verbose` and `verbose_file` into just `verbose::Union{Bool,IOStream}`. I still think the global option would be better. To fix the issue you were having to ensure you close the `FILE` you would just need to keep a reference to an `IOStream` and keep doing what you are doing.

File: src/FTPC.jl (hunk 2 above)
Comment (samuel-massinon) [inline src/FTPC.jl]: Didn't work. Works locally but only by luck.

Comment (omus) [inline src/FTPC.jl]: Ok. What you have is fine

**Comment by samuel-massinon** [comment]
@omus By global reference do you mean a variable for the module or for the object? Or do you mean that users can set verbose to on without have to pass `verbose=true` all the time?

**Comment by omus** [comment]
I originally was thinking more along the lines of a variable at the module level but I think keeping the verbose level associated with an object may make more sense. The current interface is rather difficult to work with when you want to turn on verbose for all of the FTP commands. Maybe the solution is to use something like https://github.com/WestleyArgentum/Lumberjack.jl?

**Comment by samuel-massinon** [comment]
How about keep the current interface but also include a `verbose` variable in the object. If either are true or is an `IOStream`, then set verbose to true? This way a user can set it from the start, or use it once in a while.

**Review by omus — COMMENTED** [review]
Nothing glaring popped out. The CI issues look relevant to this PR.

File: src/FTPObject.jl [hunk 4]
```diff
@@ -154,7 +208,8 @@ function upload(
     local_file_paths::Vector{<:AbstractString},
     ftp_dir::AbstractString;
     retry_callback::Function=(count, options) -> (count < 4, options),
-    retry_wait_seconds::Integer = 5
+    retry_wait_seconds::Integer = 5,
+    verbose::Union{Bool,IOStream}=false,
```
Comment (omus) [inline src/FTPObject.jl]: You should make the spacing around the equals the same

File: test/ftp_object.jl [hunk 5]
```diff
@@ -389,10 +389,259 @@ end
...
+
+                str = "ABC\n"
+                write(io, str)
+
+                # @test position(io) == first_pos + length(str)
+                println("\n\n")
+                path = pwd(ftp; verbose=true)
+                second_pos = position(io)
+
+                # @test second_pos == first_pos * 2 + length(str)
```
Comment (omus) [inline test/ftp_object.jl]: Did you for get to uncomment these?

**Review by omus — APPROVED** [review]
Looks good to me. I think we should squash this since there are so many changes...

File: test/ftp_object.jl (hunk 5 above)
Comment (kano31) [reply test/ftp_object.jl]: No that was on purpose, I wanted to see the output since the tests were failing.

**Pull request merged**